### PR TITLE
Use whoosh in DB

### PR DESCRIPTION
**Pull request by justcalamari**

### CODE REVIEW

File: libcflib/db.xsh [hunk 1]
```diff
@@ -49,11 +51,9 @@ class DB:
         self._channel_graphs = {}
         self._packages = {}
         self._initialized = True
+        self._idx = $LIBCFGRAPH_INDEX
 
-    def _build_whoosh(self):
-        self.idx = 'whoosh'
-
-    def search(self, **kwargs):
+    def search(self, query):
```
Comment (CJ-Wright) [inline libcflib/db.xsh]: Would it be possible to have some example queries in the docs?

**Comment by scopatz** [comment]
Thanks for putting this in @justcalamari! Also, would you mind adding a test, please? Thanks!

File: libcflib/db.xsh [hunk 2]
```diff
@@ -43,7 +43,7 @@ class DB:
                 git pull $LIBCFGRAPH_URL master
         else:
             LOGGER.log('grabbing initial graph', category="db")
-            git clone --quiet $LIBCFGRAPH_URL $LIBCFGRAPH_DIR
+            git clone --depth=1 --quiet $LIBCFGRAPH_URL $LIBCFGRAPH_DIR
```
Comment (CJ-Wright) [inline libcflib/db.xsh]: I think this caused problems in the past. I don't remember what they were though.

Comment (scopatz) [reply libcflib/db.xsh]: This change was requested by @mariusvniekerk

Comment (CJ-Wright) [reply libcflib/db.xsh]: Ah ok, then I guess it isn't a problem then!

Comment (scopatz) [reply libcflib/db.xsh]: Ahh, I mean that @mariusvniekerk asked for the removal of `--depth=1`

**Comment by justcalamari** [comment]
Ready for review @scopatz, @CJ-Wright, @mariusvniekerk. 

Note that we also need to add indexer to the circle config in libcfgraph, but we can remove it after we finish indexing all the artifacts already in there.

I also need to add news before merging.

File: libcflib/db.xsh [hunk 3]
```diff
@@ -49,17 +51,15 @@ class DB:
...
-
-    def search(self, **kwargs):
+    def search(self, query):
         """Search the database
 
         Parameters
         ----------
-        kwargs : dict
-            The keys to search on
+        query: dict
```
Comment (CJ-Wright) [inline libcflib/db.xsh]: I think this needs a space. `query : dict`

File: libcflib/db.xsh [hunk 4]
```diff
@@ -49,17 +51,15 @@ class DB:
...
+            The keys to search on. e.g. `{"about.conda_version": "4.2.13"}`
```
Comment (CJ-Wright) [inline libcflib/db.xsh]: Does this need double '`'?

Comment (justcalamari) [reply libcflib/db.xsh]: Does it? I thought that single backticks were handled correctly, though I'm not sure.

**Comment by justcalamari** [comment]
Once we start adding feedstocks we won't need to reindex the artifacts unless we change the artifact files.

File: libcflib/db.xsh (hunk 4 above)
Comment (CJ-Wright) [reply libcflib/db.xsh]: Im not certain either, maybe someone else can help with this. I thought RST was two but idk if this is true RST.

**Comment by mariusvniekerk** [comment]
How long does it take to index?

**Comment by mariusvniekerk** [comment]
@justcalamari we get new artifact files every hour.

File: libcflib/indexer.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    schema = create_whoosh_schema(SCHEMAS["artifact"]["schema"])
+    schema.add("pkg", TEXT(stored=True))
+    schema.add("channel", TEXT(stored=True))
+    schema.add("arch", TEXT(stored=True))
+    schema.add("name", TEXT(stored=True))
+    schema.add("path", ID(stored=True, unique=True))
+    ix = get_index(ind, schema=schema)
+
+    unindexed = unindexed_artifacts(path, ix)
+    print(f"TOTAL UNINDEXED ARTIFACTS: {len(unindexed)}")
```
Comment (scopatz) [inline libcflib/indexer.py]: Should this be logged rather than printed?

File: libcflib/indexer.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+from concurrent.futures import as_completed, ThreadPoolExecutor
+from whoosh.fields import ID, TEXT
+
+import tqdm
+
+from .schemas import SCHEMAS
+from .whoosh.utils import create_whoosh_schema, get_index
+
+
+def all_artifacts(root):
```
Comment (scopatz) [inline libcflib/indexer.py]: It seems like all of the functions in this module need docstrings.

**Comment by mariusvniekerk** [comment]
That's pretty rough given that we have 100k 🤔

File: libcflib/whoosh/writing.py [hunk 7]
```diff
@@ -29,7 +31,7 @@ def get_value(fields, fieldname):
     return value
 
 
-def get_fieldnames(fields, base=""):
+def get_fieldnames(fields, base=[]):
```
Comment (scopatz) [inline libcflib/whoosh/writing.py]: You should not have mutable values as a default argument.  Pleas change this to an empty tuple or None, if needed.

**Comment by scopatz** [comment]
Thanks!

**Pull request merged**